### PR TITLE
refactor: make explicit classes for data structures

### PR DESCRIPTION
**Pull request by iamogbz**
## Description

This clarifies the structure of `SnapshotFile`, `SnapshotFiles`, `SnapshotData` and `AssertionResult` used for asserting, calculating and reporting of snapshot tests

## Related Issues

N/A

## Checklist

- [x] This PR has sufficient test coverage.

## Additional Comments

Should make the session logic more explicit

### CODE REVIEW

File: tests/test_amber_data_serializer.py [hunk 1]
```diff
@@ -1,25 +0,0 @@
-import pytest
-
-from syrupy.serializers.amber import DataSerializer
-
-
-@pytest.mark.parametrize(
```
Comment (noahnu) [inline tests/test_amber_data_serializer.py]: Was this removed as redundant?

Comment (iamogbz) [reply tests/test_amber_data_serializer.py]: yes

File: tests/__snapshots__/test_amber_data_serializer.ambr [hunk 2]
```diff
@@ -1,24 +0,0 @@
-# name: test_produces_valid_object_types[0]
-  "<class 'str'>"
----
-# name: test_produces_valid_object_types[1]
-  "<class 'int'>"
----
-# name: test_produces_valid_object_types[2]
-  "<class 'tuple'>"
----
-# name: test_produces_valid_object_types[3]
```
Comment (noahnu) [inline tests/__snapshots__/test_amber_data_serializer.ambr]: we should make sure we've tested each type in the remaining test files. e.g. booleans

File: src/syrupy/session.py [hunk 3]
```diff
@@ -1,5 +1,4 @@
 import os
```
Comment (noahnu) [inline src/syrupy/session.py]: This file is getting pretty large. Can we movem the data structures into a separate file / split up this file somehow?

File: tests/__snapshots__/test_amber_data_serializer.ambr (hunk 2 above)
Comment (iamogbz) [reply tests/__snapshots__/test_amber_data_serializer.ambr]: Sure

File: src/syrupy/session.py (hunk 3 above)
Comment (iamogbz) [reply src/syrupy/session.py]: Done

File: src/syrupy/data.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        if not snapshot_file_to_update:
+            snapshot_file_to_update = SnapshotFile(filepath=snapshot_file.filepath)
+            self.add(snapshot_file_to_update)
+        snapshot_file_to_update.snapshots.update(snapshot_file.snapshots)
+
+    def __iter__(self) -> Iterator["SnapshotFile"]:
+        return iter(self._snapshot_files)
+
+    def __contains__(self, key: str) -> bool:
+        return key in (snapshot_file.filepath for snapshot_file in self._snapshot_files)
```
Comment (noahnu) [inline src/syrupy/data.py]: (nb) equivalent:
```
any(key == snapshot_file.filepath for ...)
```

File: src/syrupy/serializers/raw_single.py [hunk 5]
```diff
@@ -44,17 +51,17 @@ def _read_file(self, filepath: str) -> Any:
...
-    def delete_snapshots_from_file(self, snapshot_file: str, _: Set[str]) -> None:
-        os.remove(snapshot_file)
+    def delete_snapshots_from_file(self, snapshot_filepath: str, _: Set[str]) -> None:
+        os.remove(snapshot_filepath)
 
-    def __write_file(self, filepath: str, data: "SerializableData") -> None:
-        with open(filepath, "wb") as f:
-            f.write(self.serialize(data))
+    def __write_file(self, filepath: str, data: Optional["SerializedData"]) -> None:
+        if isinstance(data, bytes):
```
Comment (noahnu) [inline src/syrupy/serializers/raw_single.py]: we should fail if data is not of type bytes assuming we can't handle it

File: src/syrupy/assertion.py [hunk 6]
```diff
@@ -23,13 +28,24 @@
     from .location import TestLocation
     from .serializers.base import AbstractSnapshotSerializer
     from .session import SnapshotSession
-    from .types import SerializableData, SnapshotFiles
+    from .types import SerializableData, SerializedData  # noqa: F401
```
[the remark this answers is not in the document]
Comment (iamogbz) [reply src/syrupy/assertion.py]: It keeps bugging out, haven't investigated why

**Pull request merged**